initial proposal of Oscar group characters

ThomasBreuer opened the pull request:
(comments are inside `GroupCharacters.jl`)

ThomasBreuer committed:
initial proposal of Oscar group characters

(comments are inside `GroupCharacters.jl`)

fingolfin commented on examples/GroupCharacters.jl:
```diff
@@ -0,0 +1,350 @@
+    GAPGroup::Oscar.GAPGroup    # the underlying group
+    GAPTable::GAP.GapObj  # the character table object
+    characteristic::Int
+end
```
Do we really need two types? We could just use one, with different inner constructor
```suggestion
struct GAPGroupCharacterTable <: GroupCharacterTable
    GAPGroup::Oscar.GAPGroup    # the underlying group, if any
    GAPTable::GAP.GapObj  # the character table object
    characteristic::Int
    function GAPGroupCharacterTable(G::Oscar.GAPGroup, tab::GAP.GapObj, char::Int)
      ct = new()
      ct.GAPGroup = G
      ct.GAPTable = tab
      ct.characteristic = char
    end
    function GAPGroupCharacterTable(tab::GAP.GapObj, char::Int)
      ct = new()
      #ct.GAPGroup is left undefined
      ct.GAPTable = tab
      characteristic = char
    end
end
```
... and then use `isdefined(ct, :GAPGroup)` to test whether a group was set.

Moreover, if this is changed to a `mutable struct`, we could even set a group later on.

ThomasBreuer replied:
O.k., I will define just one type.

I would regard the possibility to set a group later on as asking for trouble.
For example, consider two character tables without stored groups, which belong to a group and a subgroup, respectively. If somebody later sets (independently) incompatible groups, suddenly some calls will fail that worked before setting the groups.
(The worst case will then be that the inconsistency is not evident enough for running into errors, and one gets wrong results.)
In GAP, we had run into this problem, and the operation `CharacterTableWithStoredGroup` was introduced (which creates a new table) instead of the possibility to simply set a group.

fingolfin commented on examples/GroupCharacters.jl:
```diff
@@ -0,0 +1,350 @@
+  The next instance is the connection from the group to its character table.
+  If one wants something like `trivial_character(G)` then the group must
+  know how to fetch the same character table object in each call;
+  perhaps we do not want to support such functions at all.
```
I think a group should indeed store "its" character table. The easiest way to do this for now would be to add a `GAPGroupCharacterTable` entry to each concrete GAPGroup subtype (which by default would be left undefined).

More flexible would be to use the `@declare_other` in those types; this adds `Dict{Symbol,Any}` to them, which can be used to store attributes, basically exactly as in GAP component objects. Then one can use `get_special` / `set_special` to access those (see the code in https://github.com/Nemocas/AbstractAlgebra.jl/blob/171944b8c6262d5928316a7bc0ecef41258facf8/src/AbstractAlgebra.jl#L146 and following; see e.g. https://github.com/oscar-system/Oscar.jl/blob/master/src/Modules/FreeModules-graded.jl for a usage example).

This system is still rather bare, and apparently also there are some concerns about its performance; but I think we should just use that now and then refine it as needed.

ThomasBreuer replied:
Thanks for the hint. I will use the `@declare_other` approach.

fieker commented:
On Fri, Jul 10, 2020 at 02:34:52AM -0700, ThomasBreuer wrote:
> @ThomasBreuer commented on this pull request.
> 
> 
> 
> > +
> +- What is the recommended way to convert those `QabElem` or `nf_elem` values
> +  that lie in the base field to integers (or rationals)?
> +  For example, it makes sense that `degree(chi)` returns an integer
> +  (and not a `Qab_elem`) if `chi` is a (virtual) character.
> +
> +- What is the recommended way to implement caching in Julia?
> +  The most critical point is that a character table of a group
> +  relies on a fixed ordering of conjugacy classes.
> +  (O.k., this is important enough such that we should provide a field
> +  in the table struct hat holds the array of conjugacy classes.)
> +
> +  The next instance is the connection from the group to its character table.
> +  If one wants something like `trivial_character(G)` then the group must
> +  know how to fetch the same character table object in each call;
> +  perhaps we do not want to support such functions at all.
> 
> Thanks for the hint. I will use the `@declare_other` approach.
> 

To comment on some other parts:
 - in characters, you can use QabElem - this is closely related to waht
   Magma does (a sparse cyclo per character, but all cyclos are
   compatible, so defacto you're in Qab
 - character_field to me should be just an AnticNumberField, possibly
   with an invisible embedding into Qab, but possibly without.
   There is no type for abelian field, nor do I know about algorithmic
   advantages...
 - we should possibly re-visit Qab - it was a rushed job to support
   Clara's project on binomial ideals. Then it should migrate into Oscar
   (or Hecke) proper...

Otherwise: good start   
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/oscar-system/Oscar.jl/pull/124#discussion_r452735028

ThomasBreuer committed:
next iteration, according to the comments

- only one type of character tables (with or without group)

- store/cache attributes in groups and character tables using `@declare_other`, `get_special`, `set_special`;
  for that, changed the `struct`s that represent groups and character tables to `mutable struct`s

fingolfin commented on examples/GroupCharacters.jl:
```diff
@@ -0,0 +1,348 @@
+
+Base.length(tbl::GAPGroupCharacterTable) = return length(GAP.Globals.Irr(tbl.GAPTable))
+AbstractAlgebra.nrows(tbl::GAPGroupCharacterTable) = return length(GAP.Globals.Irr(tbl.GAPTable))
+AbstractAlgebra.ncols(tbl::GAPGroupCharacterTable) = return length(GAP.Globals.Irr(tbl.GAPTable))
```
```suggestion
Base.length(tbl::GAPGroupCharacterTable) = length(GAP.Globals.Irr(tbl.GAPTable))
AbstractAlgebra.nrows(tbl::GAPGroupCharacterTable) = length(GAP.Globals.Irr(tbl.GAPTable))
AbstractAlgebra.ncols(tbl::GAPGroupCharacterTable) = length(GAP.Globals.Irr(tbl.GAPTable))
```

Naive question: Is the "length" of a character table a standard notion? Personally (as a non-expert) I'd rather think of the "size" or perhaps the number of rows and columns (which nicely matches with those nrows/ncrols methods).

Ahh, perhaps this is required for getindex to work?

ThomasBreuer replied:
Now I think that introducing `length` for character tables would be confusing in Julia: For a matrix, `length` returns the number of entries.
I will remove this method.

fingolfin commented on examples/GroupCharacters.jl:
```diff
@@ -0,0 +1,348 @@
+##
+##  characters
+##
+abstract type GroupClassFunction end
```
Is there any other notion of "class function" that we might clash with, and that would justify adding this "Group" prefix? Google and Wikipedia inform me that there is a notion of "class function" in abstract set theory (a function whose domain is not a set but a proper class), but it seems unlikely this will ever matter to us?

ThomasBreuer replied:
From my point of view, class functions are a generalization of (generalized) characters, and the term character occurs in several areas. (For example, there are character tables of endomorphism rings.) Thus it is appropriate to talk about group character objects in the given context, and then would be strange to coin the general term class function.

fingolfin commented on examples/GroupCharacters.jl:
```diff
@@ -0,0 +1,348 @@
+#function trivial_character(G::Oscar.GAPGroup)
+#end
+
+Base.length(chi::GAPGroupClassFunction) = return length(chi.values)
```
```suggestion
Base.length(chi::GAPGroupClassFunction) = length(chi.values)
```

fingolfin commented on examples/GroupCharacters.jl:
```diff
@@ -0,0 +1,348 @@
+Base.one(chi::GAPGroupClassFunction) = trivial_character(chi.table)
+
+function scalar_product(chi::GAPGroupClassFunction, psi::GAPGroupClassFunction)
+    chi.table === psi.table || error("the two class functions belong to different character tables")
```
This error message differs from the others above?

fingolfin reviewed (approved):
Looks sensible to me. Perhaps @fieker has more feedback, esp. on the cyclotomics vs. Qab bits?

ThomasBreuer committed:
character table display in Julia

- general function for displaying matrices together with row and column labels (based on the ideas of GAP's `Browse`)
- code for the special case of character tables

ThomasBreuer committed:
dealt with some suggestions by @fingolfin

- removed `length` for character tables (was a bad idea)
- made character tables and class functions iterable
- made some cosmetic changes

ThomasBreuer committed:
next iteration for the display of labelled matrices

- support the distribution into column and row portions
- support separating horizontal and vertical lines
- in the case of character tables:
  support symbolic irrationalities in the table,
  which are explained in a legend (in the footer)

fingolfin commented on examples/GroupCharacters.jl:
```diff
@@ -0,0 +1,558 @@
+end
+
+function character_table(id::String, p::Int = 0)
+    if GAP.Globals.IsBoundGlobal(GAP.julia_to_gap("CTblLib"))
```
I would recommend to concentrate input validation and availability checks at the start of the function, i.e., exit early if some constraint is not satisfied. That reduces the need for code indentation and IMHO also makes it easier to read and understand the code. So something like this, perhaps:
```suggestion
    GAP.Globals.IsBoundGlobal(GAP.julia_to_gap("CTblLib")) || error("no character table library available")
```
or
```suggestion
    hasproperty(GAP.Globals, :CTblLib) || error("no character table library available")
```

fingolfin reviewed (approved):
This is really nice. I suggest we merge it now and then iterate on it later. @ThomasBreuer is that OK by you? If so, either merge it directly or let me know and I'll merge it.

I left one comment but that can be commented in a future PR or simply ignored for now, it's a minor stylistic comment.
